spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java

```diff
@@ -64,28 +75,56 @@ public AppRegistration find(String name, ApplicationType type) {
 	public List<AppRegistration> findAll() {
 		List<AppRegistration> apps = new ArrayList<>();
 		for (Map.Entry<String, URI> entry : this.uriRegistry.findAll().entrySet()) {
-			apps.add(createAppRegistration(entry.getKey(), entry.getValue()));
+			String[] tokens = entry.getKey().split("\\.");
+			if (tokens.length == 2) {
+				String name = tokens[1];
+				ApplicationType type = ApplicationType.valueOf(tokens[0]);
+				URI metadataUri = null;
+				try {
+					metadataUri = this.uriRegistry.find(entry.getKey() + "." + METADATA_KEY_SUFFIX);
+				}
+				catch (IllegalArgumentException ignored) {
+				}
+				apps.add(new AppRegistration(name, type, entry.getValue(), metadataUri, this.resourceLoader));
+			} else {
+				Assert.isTrue(tokens.length == 3
+						&& METADATA_KEY_SUFFIX.equals(tokens[2]),
+					"Invalid format for app key '" + entry.getKey() + "'in registry. Must be <type>.<name> or <type>.<name>.metadata");
+			}
 		}
 		return apps;
 	}
 
-	public AppRegistration save(String name, ApplicationType type, URI uri) {
+	public AppRegistration save(String name, ApplicationType type, URI uri, URI metadataUri) {
 		this.uriRegistry.register(key(name, type), uri);
-		return new AppRegistration(name, type, uri, this.resourceLoader);
+		if (metadataUri != null) {
+			this.uriRegistry.register(metadataKey(name, type), metadataUri);
+		}
+		return new AppRegistration(name, type, uri, metadataUri, this.resourceLoader);
 	}
 
-	public List<AppRegistration> importAll(boolean overwrite, String... resourceUris) {
+	public List<AppRegistration> importAll(boolean overwrite, Resource... resources) {
 		List<AppRegistration> apps = new ArrayList<>();
-		for (String uri : resourceUris) {
+		for (Resource resource :resources) {
 			try {
-				Map<String, URI> registered = this.uriRegistryPopulator.populateRegistry(
-						overwrite, this.uriRegistry, uri);
+				Map<String, URI> registered = UriRegistryPopulator.populateRegistry(
+					overwrite, this.uriRegistry, resource);
 				for (Map.Entry<String, URI> entry : registered.entrySet()) {
```

Comment:
True. Not sure there is a way to do it earlier without duplicating logic of validation or shifting knowledge about key format etc. Ideas?